server.js

```diff
@@ -731,166 +731,127 @@ app.post('/submit-match',  asyncMiddleware( async (req, res, next) => {
+
+    // match is found in pending_matches
+    if (pending_match_index >= 0) {
+        if (sprt_result === false) {
+            // remove from pending matches
+            console.log("SPRT: Early fail pop: " + JSON.stringify(pending_matches.splice(pending_match_index, 1)));
+            console.log("SPRT: Early fail post-pop: " + JSON.stringify(pending_matches));
+        } else {
+            var m = pending_matches[pending_match_index];
+
+            // remove the match from the requests array.
+            var index = m.requests.findIndex(e => e.seed === seed_from_mongolong(req.body.random_seed));
+            if (index !== -1) {
+                m.requests.splice(index, 1);
+            }
+
+            // update stats 
+            m.game_count++;
+            if (m.network1 == req.body.winnerhash) {
+                m.network1_wins++;
+            } else {
+                m.network1_losses++;
+            }
+
+            // Check > 1 since we'll run to 400 even on a SPRT pass, but will do it at end.
+            //
+            if (sprt_result === true && pending_matches.length > 1) {
+                console.log("SPRT: Early pass unshift: " + JSON.stringify(m));
+                pending_matches.unshift(pending_matches.splice(pending_match_index, 1));
+            }
+        }
+    }
+
+    // Lastly, promotion check!!
     // Check if network2 == best_network_hash and if so, check SPRT. If SPRT pass, promote network1 as new best-network.
     // This is for the case where a match comes in to promote us, after it is no longer the active match in queue.
     //
-    if (!new_best_network_flag && req.body.loserhash == best_network_hash) {
-        db.collection("matches").findOne({ network1: req.body.winnerhash, network2: best_network_hash, options_hash: req.body.options_hash})
-            .then((match) => {
-            // added !is_test flag
-            //
-            if (match && !match.is_test && ( (SPRT(match.network1_wins, match.network1_losses) === true) || (match.game_count >= 400 && match.network1_wins / match.game_count >= 0.55) ) ) {
-                fs.copyFileSync(__dirname + '/network/' + req.body.winnerhash + '.gz', __dirname + '/network/best-network.gz');
-                console.log("New best network copied from (normal check): " + __dirname + '/network/' + req.body.winnerhash + '.gz');
-            }
-        }).catch( err => {
-            console.log("ERROR: " + req.body.winnerhash + " " + best_network_hash + " " + req.body.options_hash);
-            console.log("ERROR: Couldn't check for new best network: " + err);
-        });
+    if (
+        // Best network has lost
+        req.body.loserhash == best_network_hash
+        // This is not a test match
+        && !match.is_test
+        // SPRT passed OR it has reach 55% after `number_to_play` games
+        && (
+            sprt_result === true
+            || (match.game_count >= match.number_to_play && match.network1_wins / match.game_count >= 0.55)
```

Comment:
Should we hard code this test as >= 400 in case we queue a match with < 400 number_to_play? We've never done that, so I'm not sure. But if we did queue for less, say 100 games, would we want to promote a 55% with 101 games using the simple 55% check without the SPRT passing?

Reply:
It looks like SPRT function has magic number 400 hard coded, then it’s better to use 400 here too